include/seqan3/alignment/pairwise/alignment_configurator.hpp

```diff
@@ -0,0 +1,189 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::alignment_selector.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <functional>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include <seqan3/alignment/configuration/all.hpp>
+#include <seqan3/alignment/pairwise/policy/all.hpp>
+#include <seqan3/alignment/pairwise/alignment_algorithm.hpp>
+#include <seqan3/alignment/pairwise/align_result.hpp>
+#include <seqan3/alignment/pairwise/edit_distance_unbanded.hpp>
+#include <seqan3/alignment/pairwise/select_result_type.hpp>
+#include <seqan3/alphabet/gap/gapped.hpp>
+#include <seqan3/core/concept/tuple.hpp>
+#include <seqan3/core/metafunction/deferred_crtp_base.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/core/metafunction/template_inspection.hpp>
+#include <seqan3/core/type_list.hpp>
+#include <seqan3/range/view/persist.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief Provides several contracts to test when configuring the alignment algorithm.
+ * \ingroup pairwise_alignment
+ * \tparam range_type            The type of the range containing sequences to be aligned.
+ * \tparam alignment_config_type The type of the alignment configuration.
+ *
+ * \details
+ *
+ * This stateless helper class provides several contract testing functions for the alignment configuration.
+ */
+template <typename range_type,
+          typename alignment_config_type>
+struct alignment_contract
+{
+private:
+    /*!\brief Auxiliary member types
+     * \{
+     */
+    using first_seq_t  = std::tuple_element_t<0, value_type_t<std::ranges::iterator_t<range_type>>>;
+    using second_seq_t = std::tuple_element_t<1, value_type_t<std::ranges::iterator_t<range_type>>>;
+    //!\}
+
+public:
+    //!\brief Tests whether the value type of `range_type` is a tuple with exactly 2 members.
```

Comment:
As said before, this could be an std::pair. In that case this test would be omitted.

Reply:
We cannot assume anything about the type here. So I cannot require that sequences always have to be passed as std::pair. For example using the ranges::view::zip (which is very likely in this context) will return a tuple not a pair.